### PR TITLE
feat(contacts): Make our Contacts vCard-backed

### PR DESCRIPTION
**Pull request by tadzik**
Includes a switch to a new (hopefully better) API

Issues discovered:

- [x] [Select multiple should allow for adding to groups](https://github.com/runbox/runbox7/pull/554#issuecomment-627423054) -- fixed in f46f6f4
- [x] [Drag and drop doesn't work on Vivaldi](https://github.com/runbox/runbox7/pull/554#issuecomment-6274230540) -- "fixed" in 4271477 -- added an optional drag handle
- [x] [Save changes does nothing on contacts with no name](https://github.com/runbox/runbox7/pull/554#issuecomment-627423054) -- fixed in dbd5ab029a81ac4c094645bbb2baafecbff36316
- [x] [Disallow recursive groups](https://github.com/runbox/runbox7/pull/554#issuecomment-627423054) -- partially fixed by 60bfc89
- [x] [VCF property types not parsed correctly](https://github.com/runbox/runbox7/pull/554#issuecomment-627423054)
- [x] [VCF group syntax not parsed (or stripped...) correctly](https://github.com/runbox/runbox7/pull/554#issuecomment-627423054) -- "fixed" in bf2f951
- [x] Needs a way to email a group of contacts -- fixed in f287c87
- [x] Review use of localstorage? https://dev.to/rdegges/please-stop-using-local-storage-1i04 -- migrated to its own task: https://github.com/runbox/runbox7/issues/585
- [x] [Importing a vCard with no names (only email) - contact list shows no info](https://github.com/runbox/runbox7/pull/554#issuecomment-628560966) -- fixed in a96fd92
- [x] Importing a vCard with a UUID matching an existing one - Internal server error (from REST PUTing to carddav) -- fixed in https://github.com/runbox/runbox7/pull/554/commits/2c504972af0a56a828c1453e48642f680065caf9
- [x] When importing vcf, allow user to choose whether to keep/add the "categories" from the cards in the file -- fixed in e72ba4c
- [x]  [cache overflow error](https://github.com/runbox/runbox7/pull/554#issuecomment-630241759) -- should be fixed by 8c19f68
- [x] setting birthdays on contacts doesn't seem to work (tested on "22/05/1971") -- "fixed" in 3d5415a
- [x] - UI (both drag & drop and select multiple) allows me to add, and save a contact into a group thats already in that group. -- fixed in efa6808
- [ ] Birthday 500ing or doesnt round trip? I saved "09-21", which gave a 500-error (tried to save as BDAY;VALUE=DATE:09-21), the 2nd attempt then saved. when I reloaded it said "09--21", I hit save and that errored too (BDAY;VALUE=DATE:09--21) 
- [ ] If I type some things into Notes that should be escaped/round trip, they don't quite; "\r\n" causes the "\n" to vanish. \\ loses one of the slashes. I can't get it to break out though.

### CODE REVIEW

**Comment by gtandersen** [comment]
Looks good @tadzik! Could you have a look as well @castaway?

**Comment by castaway** [comment]
Going to add functionality thoughts/notes here so I don't forget:

* Idea: "select multiple" (which seems to be for multi-select/delete) could also pop up an "add all icon" if used when group editing is open?

* Issue: Drag&Drop (contact to group) is ultra slow on my browser:
Vivaldi	3.0.1874.33 (Stable channel) (64-bit)
Revision	ca4ce8c6f5fd2c35122081da7ed4a8533d05323d
OS	Linux
JavaScript	V8 8.1.307.31
Flash	32.0.0.293 /home/castaway/.config/google-chrome/PepperFlash/32.0.0.293/libpepflashplayer.so
User Agent	Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.132 Safari/537.36
Command Line	/usr/bin/vivaldi --flag-switches-begin --flag-switches-end --disable-webrtc-apm-in-audio-service --save-page-as-mhtml
Executable Path	/opt/vivaldi/vivaldi
Profile Path	/home/castaway/.config/vivaldi/Default

* "Save changes" on a group with no name does.. nothing at all, visually - I think it needs an error message box with "Please add a name to save the new group" or similar

File: src/app/contacts-app/contact.spec.ts [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    });
+
+    it('can parse ADR property', () => {
+        const vcard = `BEGIN:VCARD
+VERSION:3.0
+PRODID:-//Sabre//Sabre VObject 4.1.6//EN
+UID:3e65dced-9784-49ea-b9d7-9aac66c00f99
+REV;VALUE=DATE-AND-OR-TIME:20200424T110821Z
+FN:Addressman
+ADR;TYPE=HOME:pobox 123;;foo street 13;Townsville;Centralia;12-345;Contactia
```
Comment (tadzik) [inline src/app/contacts-app/contact.spec.ts:91]: This test isn't checking if `type` is set correctly (and apparently it isn't)

**Comment by castaway** [comment]
Fixed the "item1." issue over here: https://github.com/mozilla-comm/ical.js/pull/442

**Comment by castaway** [comment]
image


Last item is after importing a minimal .vcf (exported from FastMail) - 
```
BEGIN:VCARD
VERSION:3.0
FN: 
N:;;;;
NICKNAME:
ORG:;
TITLE:
EMAIL;TYPE=home;TYPE=pref:jrobinson@test.org
NOTE:
PRODID:-//CyrusIMAP.org//Cyrus 3.1.7-238-g170a812-fmstable-20190913v1//EN
REV:20190917T153944Z
UID:145171b7-ba96-4694-963d-c4c180c02a00
END:VCARD
```

on prod this displays the email in the list.

**Comment by castaway** [comment]
New fun, storage overflow error on localstorage "cacheContacts", content is now: https://gist.github.com/castaway/c0faaeec87885808a230af75652507c9

Error: Failed to  execute 'setItem' on 'Storage': Setting the value of  '5:contactsCache' exceeded the quota.

Looks like the quota is 5MB, and I shouldnt have hit it yet - one has a URL which has expanded to fill the space available!

**Comment by tadzik** [comment]
Fixes GH-602

**Comment by tadzik** [comment]
I wasn't able to reproduce the remaining two issues – birthdays and notes. Let me know if you have some good steps to reproduce (or a testcase :)) that can show how it fails.

**Comment by castaway** [comment]
> I wasn't able to reproduce the remaining two issues – birthdays and notes. Let me know if you have some good steps to reproduce (or a testcase :)) that can show how it fails.

For notes I imported this:
```
NOTE:This is just a test for vCards.\n
 \\r\\n
 CRLF\n
 \\\\ some slashes\n
```

birthdays seems to be fixed (or at least it no longer has a 500-error on rmmstage02)

**Comment by tadzik** [comment]
A'ight. I assume that everything's acceptably working, or at least not obviously breakable. Will merge this when things turn green.

**Pull request merged**